hal/src/nRF52840/ble_hal.cpp

```
@@ -15,739 +15,3147 @@
+    return SYSTEM_ERROR_NONE;
+}
+
+void BleObject::Broadcaster::toPlatformAdvParams(ble_gap_adv_params_t* params, const hal_ble_adv_params_t* halParams) {
+    memset(params, 0x00, sizeof(ble_gap_adv_params_t));
+    params->properties.type = halParams->type;
+    params->properties.include_tx_power = false; // FIXME: for extended advertising packet
+    params->p_peer_addr = nullptr;
+    params->interval = halParams->interval;
+    params->duration = halParams->timeout;
+    params->filter_policy = halParams->filter_policy;
+    params->primary_phy = BLE_GAP_PHY_1MBPS;
+}
+
+void BleObject::Broadcaster::toPlatformAdvData(ble_gap_adv_data_t* data) {
+    data->adv_data.p_data = advData_;
+    data->adv_data.len = advDataLen_;
+    data->scan_rsp_data.p_data = scanRespData_;
+    data->scan_rsp_data.len = scanRespDataLen_;
+}
+
+int BleObject::Broadcaster::configure(const hal_ble_adv_params_t* params) {
+    int ret;
+    ble_gap_adv_data_t bleGapAdvData;
+    toPlatformAdvData(&bleGapAdvData);
+    if (params == nullptr) {
+        ret = sd_ble_gap_adv_set_configure(&advHandle_, &bleGapAdvData, nullptr);
+    } else {
+        ble_gap_adv_params_t bleGapAdvParams;
+        toPlatformAdvParams(&bleGapAdvParams, params);
+        LOG_DEBUG(TRACE, "BLE advertising interval: %.3fms, timeout: %dms.",
+                  bleGapAdvParams.interval*0.625, bleGapAdvParams.duration*10);
+        ret = sd_ble_gap_adv_set_configure(&advHandle_, &bleGapAdvData, &bleGapAdvParams);
     }
-    LOG_DEBUG(TRACE, "RAM start: 0x%08x", (unsigned)ramStart);
-    // Enable the stack
-    ret = nrf_sdh_ble_enable(&ramStart);
-    if (ret != NRF_SUCCESS) {
-        LOG(ERROR, "nrf_sdh_ble_enable() failed: %u", (unsigned)ret);
-        return halError(ret);
+    NRF_CHECK_RETURN(ret, "sd_ble_gap_adv_set_configure");
+    return SYSTEM_ERROR_NONE;
+}
+
+int8_t BleObject::Broadcaster::roundTxPower(int8_t value) {
+    uint8_t i = 0;
+    for (i = 0; i < sizeof(validTxPower_); i++) {
+     
```

Errors are not checked.

---

>BleObject::init()
>dispatcher_->init();

Done.